zkevm-circuits/src/taiko_pi_circuit_.rs

```diff
@@ -0,0 +1,832 @@
+                for (annotation, block_number, bytes) in assignments {
+                    println!("{:?} {:?}, len {:?} \n    {:?}", i, annotation, bytes.len(), bytes);
+                    match i {
+                        L1SIGNAL_IDX..=FIELD10_IDX => {
+                            let field = self.field_gadget.assign(&mut region, i, keccak_r, &bytes)?;
+                            block_acc = self.assign_acc(&mut region, i, keccak_mult, block_acc, field)?;
+                            
+                            // // k_cnt += N;
+                            // if i == FIELD10_IDX {
+                            //     println!("FIELD10_IDX {:?}", block_acc);
+                            // }
+
+                            if let Some(block_number) = block_number {
+                                println!("  block_number {:?} => {:?} {:?}", block_number, self.block_number[i - PARENT_HASH].expr().identifier(), field);
+                                println!("  {:?}", bytes);
+                                assign!(region, self.block_number[i - PARENT_HASH], 0 => block_number.as_u64().scalar())?;
+                            }
+                        }
+                        KECCAK_HI..=KECCAK_LOW => {
+                            let mut bytes = bytes.clone();
+                            bytes.reverse(); 
+                            let field = self.field_gadget.assign(&mut region, i, evm_word_r, &bytes)?;
+                            block_acc = self.assign_acc(&mut region, i, evm_mult, block_acc, field)?;
+                            
+                            // // e_cnt += N/2;
+                            // if i == KECCAK_LOW {
+                            //     println!("KECCAK_LOW {:?}", block_acc);
+                            // }
+                            
+                            pi_cells.push(assign!(
+                                region, self.keccak_output[i - KECCAK_HI], 0 => rlc::value(&bytes, hi_low_r)
+                            )?);
+                        },
+                        _ => unimplemented!()
+                    }
+                    println!("  block_acc {:?}", block_acc);
+                    i += 1;
+                }
+                Ok(pi_cells)
+            }
+        )?;
+        for (i, cell) in pi_cells.iter().enumerate() {
+            layouter.constrain_instance(cell.cell(), self.public_input, i)?;
+            println!("pi_cell {:?}", cell);
+        }
+        Ok(())
+    }
+
+    pub(crate) fn assign_acc(
+        &self,
+        region: &mut CachedRegion<'_, '_, F>,
+        offset: usize,
+        acc_r: F,
+        prev: F,
+        other: F,
+    ) -> Result<F, Error> {
+        let next = prev * acc_r + other;
+        assign!(region, (self.block_acc, offset) => next)?;
+        Ok(next)
+    }
+}
+
+
+#[derive(Clone, Debug)]
+struct FieldBytesGadget<F> {
+    bytes: [Cell<F>; N],
+    is_field: Cell<F>,
+    is_first: Cell<F>,
+    is_last: Cell<F>,
```

Comment:
These are advice cells, but seems like they should be fixed cells? (should not be possible for the prover to set these values).